// @pinegrow/vite-plugin/dist/vue/index.js
import {
	onBeforeMount,
	onMounted,
	onBeforeUnmount,
	getCurrentInstance,
	ref,
	reactive,
	onUpdated,
	computed,
	watch,
} from 'vue'

export function usePinegrow() {
	const initCache = () => {
		// conditional
		const winObj = window

		if (!(winObj?.process?.client && winObj.process.client !== true)) {
			if (!winObj.pinegrow) {
				// console.log('Cache initialized by Vue Plugin!')
				const elCache = reactive(new Map())

				const enrichWithComponentName = (elCacheObj, localFile) => {
					return {
						name: elCacheObj.instance.type.__name || '',
						localFile,
						...elCacheObj,
					}
				}

				winObj.pinegrow = {
					elCache,
					// pgIdToElComputed,
					// localComponentToElComputed,
					// appTree,
					reactiveFromContext: reactive,
					refFromContext: ref,
					computedFromContext: computed,
					watchFromContext: watch,
					// // Uncomment these two to test locally
					// elCacheErrHandlerFn,
					// elUpdateHanderFn,
				}
			}
		}
	}

	// pgId & key can be optional
	const pgUpdateElCache = (hook, pgId, rootEl, key, localFile) => async vnode => {
		if (!vnode) return
		if (window?.process?.client && process?.client !== true) return
		if (!window.pinegrow) initCache()

		let el = vnode.el

		const instance = vnode.component || vnode.ctx || el.__vueParentComponent

		if (!el || !instance) return

		try {
			if ((key !== null && key !== undefined) || (vnode.key !== null && vnode.key !== undefined)) {
				key = vnode.key || key
			}

			localFile = localFile || vnode.type.__file

			let isRootFragment = false
			let isFragment = el.nodeType !== 1
			let firstEl,
				lastEl,
				isIsland = false,
				childVNodes = []

			// May be an iles Island that wraps components with client directives
			if (localFile) {
				if (localFile.includes('node_modules/iles') && localFile.includes('Island.vue')) {
					// Retain localFiles of iles island to apply to child elements
					isIsland = true
				}

				if (localFile.includes('node_modules')) {
					// Ignore SFCs from node_modules
					localFile = null
				}
			}

			const localFileFromInstance = instance.type.__file
			if (
				localFileFromInstance &&
				localFileFromInstance.includes('node_modules/iles') &&
				localFileFromInstance.includes('Island.vue')
			) {
				// Retain localFiles of iles island to apply to child elements
				isIsland = true

				localFile = instance.props.component?.__file || instance.props.importFrom
			}

			// // Computed props anyway filters out unmounted ones, and we use the local components unmounted hooks to cleanup unmounted nodes. Using the unmounted hook seems to be out of sync when el is reused but instance is remounted.
			// if (hook === 'unmounted') {
			//   if (pinegrow.elCache.has(el)) {
			//     pinegrow.elCache.delete(el)
			//   }
			//   for (let [key, value] of pinegrow.elCache.entries()) {
			//     if (value.rootEl === rootEl) {
			//       pinegrow.elCache.delete(key)
			//     }
			//   }
			//   return
			// }

			// Text/comment node
			if (isFragment) {
				if (!rootEl) {
					// root Fragment
					isRootFragment = true
				}

				function isFragmentEl(instance) {
					// return appRecord.options.types.Fragment === instance.subTree?.type
					return instance.subTree.el.nodeType !== 1
				}

				function getRootVNodesFromComponentInstance(instance) {
					if (isFragmentEl(instance)) {
						return getFragmentRootVNodes(instance.subTree)
					}
					if (!instance.subTree) return []
					return [instance.subTree]
				}

				function getFragmentRootVNodes(vnode) {
					if (!vnode.children) return []
					// children is v-if when the vnode has a condition
					if (!Array.isArray(vnode.children)) return []

					const list = []

					for (let i = 0, l = vnode.children.length; i < l; i++) {
						const childVnode = vnode.children[i]
						if (childVnode.component) {
							list.push(...getRootVNodesFromComponentInstance(childVnode.component))
						} else if (childVnode) {
							list.push(childVnode)
						}
					}

					return list
				}

				childVNodes = getRootVNodesFromComponentInstance(instance)

				if (!childVNodes.length) {
					// For NuxtLayout, no childVNodes are returned, the subTree.children is in fact an object {ctx: {}, default: f}
					if (el.nextElementSibling) {
						const childEl = el.nextElementSibling
						firstEl = lastEl = childEl
						const childInstance = childEl.$ || childEl.__vueParentComponent
						const childVNode = childInstance?.vnode // or subTree?
						if (childVNode) {
							childVNodes.push(childVNode)
						}
					}
				}

				if (childVNodes.length) {
					// Filter out recursive vnodes text1 -> div, text1 (happens when text1 is the closest to a fragment with div & slot)
					childVNodes = childVNodes.filter(childVNode => childVNode.el !== el)

					if (childVNodes.length) {
						if (childVNodes.length === 1) {
							firstEl = lastEl = childVNodes[0].el
						} else {
							firstEl = childVNodes[0].el
							lastEl = childVNodes[childVNodes.length - 1].el
						}

						if (firstEl && firstEl.nodeType !== 1) {
							firstEl = firstEl.nextElementSibling
						}
						if (lastEl && lastEl.nodeType !== 1) {
							lastEl = lastEl.nextElementSibling
						}
					}
				}
			}

			if (isIsland && !isRootFragment) {
				return
			}

			let elCacheObj = {
				el,
				isRootFragment,
				rootEl,
				vnode,
				instance,
				isFragment,
				firstEl,
				lastEl,
				pgId,
				key,
				localFile,
				isIsland,
			}

			let prevElCacheNodes = pinegrow.elCache.get(el)

			if (prevElCacheNodes) {
				let index = prevElCacheNodes.findIndex(elCacheObj => elCacheObj.instance.uid === instance.uid)

				if (index > -1) {
					const prevElCacheObj = prevElCacheNodes[index]
					elCacheObj.localFile = elCacheObj.localFile || prevElCacheObj.localFile
					if (elCacheObj.pgId && prevElCacheObj.pgId && elCacheObj.pgId !== prevElCacheObj.pgId) {
						if (!prevElCacheNodes.map(prevElCacheNode => prevElCacheNode.pgId).includes(elCacheObj.pgId)) {
							prevElCacheNodes.push(elCacheObj)
						}
					} else {
						elCacheObj.pgId = elCacheObj.pgId || prevElCacheObj.pgId
						prevElCacheNodes[index] = elCacheObj
					}
				} else {
					prevElCacheNodes.push(elCacheObj)
				}
				pinegrow.elCache.set(el, prevElCacheNodes)
			} else {
				pinegrow.elCache.set(el, [elCacheObj])
				childVNodes.forEach(childVNode => {
					pgUpdateElCache(hook, pgId, isRootFragment ? el : rootEl, key)(childVNode)
				})
			}

			if (pinegrow.elUpdateHanderFn) {
				pinegrow.elUpdateHanderFn(el)
			}
		} catch (err) {
			console.log(err)
			if (pinegrow.elCacheErrHandlerFn) {
				pinegrow.elCacheErrHandlerFn(vnode, hook, rootEl, pgId, key, el, instance, err.message)
			}
		}
	}

	const elUpdateHanderFn = el => {
		// if (pinegrow.elCache.has(el)) {
		// 	const { pgId } = pinegrow.elCache.get(el)
		// 	if (pgId) {
		// 		// console.log(`Reselect ${pgId}`)
		// 	}
		// }
	}

	const elCacheErrHandlerFn = () => {
		if (message) {
			// console.log(message)
		}
	}

	const cleanupCache = () => {
		// for (let [key, value] of pinegrow.elCache.entries()) {
		//   if (value.isFragment && !value.firstEl) {
		//     console.log(value)
		//   }
		// }
		for (let [el, elCacheNodes] of pinegrow.elCache.entries()) {
			// if (!el.isConnected) {
			// 	pinegrow.elCache.delete(el)
			// 	continue
			// }
			const cleanedUpElCacheNodes = elCacheNodes.filter(
				elCacheObj => !elCacheObj.instance.isUnmounted // || elCacheObj.localFile
			)
			if (cleanedUpElCacheNodes.length) {
				if (cleanedUpElCacheNodes.length !== elCacheNodes.length) {
					pinegrow.elCache.set(el, cleanedUpElCacheNodes)
				}
			} else {
				pinegrow.elCache.delete(el)
			}
			if (pinegrow.elUpdateHanderFn) {
				pinegrow.elUpdateHanderFn(el)
			}
		}
	}

	// Local Component
	const rootVNode = ref(null)

	const mountLocalComponent = () => {
		const instance = getCurrentInstance()
		const vnode = instance?.vnode
		const el = vnode?.el
		const localFile = instance.type.__file && !instance.type.__file.includes('node_modules') && instance.type.__file
		if (instance && vnode && el) {
			rootVNode.value = vnode
			pgUpdateElCache('mounted', null, null, null, localFile)(vnode)
		}
	}

	const unmountLocalComponent = () => {
		if (rootVNode.value) {
			pgUpdateElCache('unmounted')(rootVNode.value)
		}
		cleanupCache()
	}

	onBeforeMount(() => initCache())
	onMounted(() => {
		mountLocalComponent()
	})
	onBeforeUnmount(() => unmountLocalComponent())
	onUpdated(() => {
		mountLocalComponent()
	})

	return { pgUpdateElCache }
}
